aws-lc-rs/src/rsa/oaep.rs

```diff
@@ -0,0 +1,501 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+use core::{fmt::Debug, ptr::null_mut};
+use std::ops::Deref;
+
+use aws_lc::{
+    CBB_finish, EVP_PKEY_CTX_new, EVP_PKEY_CTX_set_rsa_mgf1_md, EVP_PKEY_CTX_set_rsa_oaep_md,
+    EVP_PKEY_CTX_set_rsa_padding, EVP_PKEY_decrypt, EVP_PKEY_decrypt_init, EVP_PKEY_encrypt,
+    EVP_PKEY_encrypt_init, EVP_PKEY_up_ref, EVP_marshal_private_key, EVP_marshal_public_key,
+    EVP_parse_public_key, EVP_sha1, EVP_sha256, EVP_sha384, EVP_sha512, EVP_MD, EVP_PKEY,
+    EVP_PKEY_CTX, RSA_PKCS1_OAEP_PADDING,
+};
+
+use crate::{
+    buffer::Buffer,
+    cbb::{self, LcCBB},
+    cbs,
+    encoding::AsDer,
+    error::{KeyRejected, Unspecified},
+    fips::indicator_check,
+    pkcs8::Document,
+    ptr::LcPtr,
+};
+
+use super::key::{generate_rsa_evp_pkey, KeySize, PKCS8_CAPACITY_BUFFER};
+
+/// RSA-OAEP with SHA1 Hash and SHA1 MGF1
+pub const OAEP_SHA1_MGF1SHA1: EncryptionAlgorithm = EncryptionAlgorithm {
+    id: EncryptionAlgorithmId::OaepSha1Mgf1sha1,
+    hash: EVP_sha1,
+    mgf: EVP_sha1,
+};
+
+/// RSA-OAEP with SHA256 Hash and SHA256 MGF1
+pub const OAEP_SHA256_MGF1SHA256: EncryptionAlgorithm = EncryptionAlgorithm {
+    id: EncryptionAlgorithmId::OaepSha256Mgf1sha256,
+    hash: EVP_sha256,
+    mgf: EVP_sha256,
+};
+
+/// RSA-OAEP with SHA384 Hash and SHA384  MGF1
+pub const OAEP_SHA384_MGF1SHA384: EncryptionAlgorithm = EncryptionAlgorithm {
+    id: EncryptionAlgorithmId::OaepSha384Mgf1sha384,
+    hash: EVP_sha384,
+    mgf: EVP_sha384,
+};
+
+/// RSA-OAEP with SHA512 Hash and SHA512 MGF1
+pub const OAEP_SHA512_MGF1SHA512: EncryptionAlgorithm = EncryptionAlgorithm {
+    id: EncryptionAlgorithmId::OaepSha512Mgf1sha512,
+    hash: EVP_sha512,
+    mgf: EVP_sha512,
+};
+
+/// RSA Encryption Algorithm Identifier
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum EncryptionAlgorithmId {
+    /// RSA-OAEP with SHA1 Hash and SHA1 MGF1
+    OaepSha1Mgf1sha1,
+
+    /// RSA-OAEP with SHA256 Hash and SHA256 MGF1
+    OaepSha256Mgf1sha256,
+
+    /// RSA-OAEP with SHA384 Hash and SHA384 MGF1
+    OaepSha384Mgf1sha384,
+
+    /// RSA-OAEP with SHA512 Hash and SHA512 MGF1
+    OaepSha512Mgf1sha512,
+}
+
+type HashFn = unsafe extern "C" fn() -> *const EVP_MD;
```

Comment:
These are both types of hash/digest functions. Would `OaepDigestFn` and `Mgf1DigestFn` be more appropriate names?